File: Stockpile.Api/Contracts/Mappers/InventoryItemMapper.cs
using Stockpile.Api.Contracts.Requests;
using Stockpile.Api.Contracts.Response;
using Stockpile.Domain.Entities;

namespace Stockpile.Api.Contracts.Mappers;

public static class InventoryItemMapper
{
    public static InventoryItem MapTo(this InventoryItemRequest request, string id, string userId, string username)
    {
        return new InventoryItem
        {
            Id = id,
            Name = request.Name,
            Quantity = request.Quantity,
            Unit = request.Unit,
            ExpiryDate = request.ExpiryDate,
            UserId = userId,
            Username = username
        };
    }

    public static InventoryItem MapTo(this InventoryItemRequest request, string userId, string username)
    {
        return new InventoryItem
        {
            Name = request.Name,
            Quantity = request.Quantity,
            Unit = request.Unit,
            ExpiryDate = request.ExpiryDate,
            UserId = userId,
            Username = username
        };
    }
    
    public static InventoryItemResponse MapFrom(this InventoryItem item)
    {
        return new InventoryItemResponse
        {
            Id = item.Id,
            Name = item.Name,
            Quantity = item.Quantity,
            Unit = item.Unit,
            ExpiryDate = item.ExpiryDate
        };
    }
}

File: smart-track-app/src/adapters/InventoryItemMapper.ts
import {
  InventoryItemRequest,
  InventoryItemResponse,
  Unit,
} from "../services/clients/api-client";
import { InventoryItem } from "../domain/models/inventory-item";

export function fromApi(dto: InventoryItemResponse): InventoryItem {
  return {
    id: dto.id,
    name: dto.name ?? "",
    quantity: dto.quantity ?? 0,
    unit: dto.unit ?? Unit.None,
    expiryDate: dto.expiryDate ? new Date(dto.expiryDate) : undefined,
  };
}

export function toApi(input: Partial<InventoryItem>): InventoryItemRequest {
  return InventoryItemRequest.fromJS({
    name: input.name,
    quantity: input.quantity,
    unit: input.unit,
    expiryDate: input.expiryDate,
  });
}
